Add email handler guide (#2072)
sensu/sensu-docs: content/sensu-go/5.16/guides/email-handler.md

---
title: "Send email alerts with the Sensu Go Email Handler"
linkTitle: "Send Email Alerts"
description: "Here’s how to send alerts to your email with the Sensu Go Email Handler. Use handlers to send events to your technology of choice (in this case, email) to alert you of incidents and help you resolve them more quickly."
weight: 85
version: "5.16"
product: "Sensu Go"
platformContent: False
menu: 
  sensu-go-5.16:
    parent: guides
---

- [Create a check](#create-a-check)
- [Add the email handler asset](#add-the-email-handler-asset)
- [Add an event filter](#add-an-event-filter)
- [Assign the email handler to a check](#assign-the-email-handler-to-a-check)
- [Create an email template](#create-an-email-template)
- [Next steps](#next-steps)

Sensu event handlers are actions the Sensu backend executes on [events][1].

> This guide explains how to use the [Sensu Go Email Handler][3] to send incident notification emails.

Reviewer: I'm a little torn on using 'incident' here. IMO it's a bit of a laden term. My suggest is to just say `send notifications emails`.

```suggestion
This guide explains how to use the [Sensu Go Email Handler][3] to send notification emails.
```

> To use this guide, you’ll need to [install the Sensu backend][12] and have at least one [Sensu agent][13] running on Linux.

Reviewer: ```suggestion
To follow this guide, you’ll need to [install the Sensu backend][12] and have at least one [Sensu agent][13] running on Linux.
```

You should also [install and configure sensuctl][4].

> ## Create a check

Reviewer: Theoretically, we don't even need to create a check--we could just create an ad-hoc event through the agent API, which might be a bit quicker than the CPU check, as folks will need to wait for CPU usage to go above the threshold or install a tool to generate load. Generating an event through the API would allow folks to receive an email quicker.

Author: Yes, I wasn't sure which path to take with this. I realized that the check-cpu check wouldn't do anything right away, so it wouldn't allow readers to get an email right away (maybe never if cpu state never changes, heh).

But going through all the steps to prepare a manually triggered event seemed like a distraction from the purpose of the guide (setting up the email handler). I think a better idea for this guide would be a check that *does* work right away -- check-cpu is just convenient because we already told them how to create it in a separate guide.

In this case, we have the Up and Running Katacoda scenario, which does include a manually triggered event that works as a live demo. I could make that clearer here...something like "if you don't want to wait for a CPU state change, try the scenario..."

Reviewer 2: I'm partial to the ad-hoc event approach because we avoid hand-waving about how to create a failing event, helping users stay focused on the email alert. We also get to remove the requirement for a running Sensu agent. 

I think using `eval $(sensuctl env)` and a curl command or two for the ad-hoc events should feel pretty straight-forward for our audience.

Instead of reference our other guide as a required step, we could move it to the conclusion, e.g. "In this guide you'll learn how to trigger an email from Sensu using ad-hoc events.", and "Now that you've completed this, the same email handler can be reused in check configurations like `check-cpu` from our `Monitor server resources` guide."

Author: Thank you both -- the updated guide uses the ad hoc event.

  namespace: default
spec:
  action: allow
  expressions:
  - event.check.occurrences == 1
  runtime_assets: []
EOF
{{< /highlight >}}

Second, create the email handler definition.
In the handler definition's `command` value, you'll need to change a few things.

Copy this text into a text editor:

{{< highlight shell >}}
cat << EOF | sensuctl create
---
api_version: core/v2
type: Handler
metadata:
  namespace: default
  name: email
spec:
  type: pipe
  command: sensu-email-handler -f YOUR-SENDER@example.com -t YOUR-RECIPIENT@example.com -s YOUR-SMTP-SERVER.example.com
    -u USERNAME -p PASSWORD
  timeout: 10
  filters:
  - is_incident
  - not_silenced
  - state_change_only
EOF
{{< /highlight >}}

Then, replace the following text:
- `YOUR-SENDER@example.com`: Replace with the email address you want to use to send email alerts
- `YOUR-RECIPIENT@example.com`: Replace with the email address you want to receive email alerts
- `YOUR-SMTP-SERVER.example.com`: Replace with the hostname of your SMTP server
- `USERNAME`: Replace with your SMTP username, typically your email address
- `PASSWORD`: Replace with your SMTP password, typically the same as your email password

> _**NOTE**: To use Gmail or G Suite as your SMTP server, follow Google's instructions for [sending email via SMTP][14]._

Reviewer: ```suggestion
_**NOTE**: To use Gmail or G Suite as your SMTP server, follow Google's instructions for [sending email via SMTP][14]. Also be aware that if you have 2fa enabled on your Google account, you'll need an [app password][15] instead of your login password._
```

> _**NOTE**: To use Gmail or G Suite as your SMTP server, follow Google's instructions for [sending email via SMTP][14]. Also be aware that if you have 2fa enabled on your Google account, you'll need an [app password][15] instead of your login password._

Author: ```suggestion
_**NOTE**: To use Gmail or G Suite as your SMTP server, follow Google's instructions for [sending email via SMTP][14]. If you have enabled 2-step verification on your Google account, you'll need to use an [app password][15] instead of your login password._
```

You probably noticed that the handler definition includes two other filters: [`is_incident`][10] and [`not_silenced`][11].
These two filters are included in every Sensu backend installation, so you don't have to create them.

After you add your email, server, username, and password values, run your updated code to create the email handler definition.

Now your handler and event filter are set up!
Next, assign the email handler to the `check-cpu` check.

## Assign the email handler to a check

With your `email` handler created, you can assign it to the `check-cpu` check.
To assign your email handler to the check `check-cpu`:

{{< highlight shell >}}
sensuctl check set-handlers check-cpu email
{{< /highlight >}}

It might take a few moments after you assign the handler to the check for the check to be scheduled on the entities and the result sent back to Sensu backend.

## Create an email template

> The last step is to create a .

Reviewer: ```suggestion
The last step is to create a template for your emails.
```

<b>Playbook</b>: https://example.com/monitoring/wiki/playbook<br>
<h3>Check Output Details</h3>
<b>Check Output</b>: {{.Check.Output}}<br>
{{end}}

<br>
#monitoringlove,<br>
<br>
Sensu<br>
</html>
{{< /highlight >}}

## Next steps

Now that you know how to apply a handler to a check and take action on events, read the [handlers reference][6] for in-depth handler documentation and check out the [Reduce alert fatigue][7] guide.

You can also follow our [Up and running with Sensu Go][9] interactive tutorial to set up the Sensu Go email handler and test it with simulated `critical` events.

[1]: ../../reference/events/
[2]: ../monitor-server-resources/
[3]: https://bonsai.sensu.io/assets/sensu/sensu-email-handler
[4]: ../../installation/install-sensu/#install-sensuctl
[5]: ../../reference/filters/
[6]: ../../reference/handlers/
[7]: ../reduce-alert-fatigue/
[8]: ../../reference/assets
[9]: ../up-running-tutorial/
[10]: ../../reference/filters/#built-in-filter-is-incident
[11]: ../../reference/filters/#built-in-filter-not-silenced
[12]: ../../installation/install-sensu/#install-the-sensu-backend
[13]: ../../installation/install-sensu/#install-sensu-agents

> [14]: https://support.google.com/a/answer/176600?hl=en

Reviewer: ```suggestion
[14]: https://support.google.com/a/answer/176600?hl=en
[15]: https://support.google.com/accounts/answer/185833?hl=en
```
